severn/api/parsers/reqfile.py

```diff
@@ -0,0 +1,150 @@
+#        contributors may be used to endorse or promote products derived from
+#        this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+__all__ = ("RequirementsFile",)
+
+import logging
+import re
+import warnings
+from pathlib import Path
+from typing import Any, List, Union
+
+import aiofiles
+
+from severn.abc import Representable
+from severn.api.constraint import Constraint
+from severn.api.dependency import Dependency
+from severn.api.utils import aenumerate
+
+ENV_MARKER_PATTERN = re.compile(r"([^<>~=!]+)(.*)")
+REQUIREMENT_PATTERN = re.compile(
+    r"""
+    (?:-r(?P<req_file>.*))?         # Requirements file
+    (?:-c(?P<con_file>.*))?         # Constraints file
+    (?:-e(?P<editable>.*))?         # Editable
+    (?P<wheel>(?:./).*)?            # Path to local distribution
+    (?P<dist_url>(?:https?://).*)?  # URL to hosted distribution
+    (?P<package>[^\[<>~=!@;]+)?     # The name of the dependency
+    (?:\[(?P<extras>.*)\])?         # Any extras the dependency specifies
+    (?:@(?P<package_url>.*))?       # The URL to a particular package's distribution
+    (?P<version>[^;]*)?             # Constraint string
+    (?:;(?P<env_markers>.*))?       # Environment markers
+    """,
+    re.VERBOSE,
+)
+
+_log = logging.getLogger(__name__)
+
+
+class RequirementsFile(Representable):
+    __slots__ = ("path",)
+
+    def __init__(self, path: Union[str, Path]) -> None:
+        self.path = path if isinstance(path, Path) else Path(path)
+
+    async def __aenter__(self) -> "RequirementsFile":
+        return self
+
+    async def __aexit__(self, *_: Any) -> None:
+        ...
+
+    async def parse(self) -> List[Dependency]:
+        dependencies: List[Dependency] = []
+
+        async with aiofiles.open(self.path) as f:
+            async for i, line in aenumerate(f, start=1):
+                line = line.strip()
+
+                if not line or line.startswith("#"):
+                    # This is quicker and more accurate than making the
+                    # regex handle it.
+                    continue
+
+                # TODO: Find a more efficient way of doing this.
+                line = line.replace(" ", "")
```

Comment:
is this meant to allow support for something like `whatever == 1.0.0` with spaces? What about something crazy like `wha te ver =     = 1       .0       .0`, we want to allow that?

Reply:
I would just as soon only support no spaces over that monstrosity lol.

Reply:
The alternative is to handle spaces in the regex, which could be done, just adds more complexity to it.

Reply:
`\s*` on both  sides of comparator wouldn't work?

Reply:
Would that only handle spaces at the start and end of the expression?

Reply:
at the start and end of the comparator. which feels like the only place they should be allowed to me, no?

Reply:
They can go [more or less anywhere](https://pip.pypa.io/en/stable/reference/requirements-file-format/).

Reply:
F in the chat. yeah just keep this replace for now we can come back to it.